pkg/querier/iterators/chunk_iterator.go

```diff
@@ -63,21 +62,71 @@ func (i *chunkIterator) At() (int64, float64) {
 }
 
 func (i *chunkIterator) AtHistogram() (int64, *histogram.Histogram) {
-	panic(errors.New("chunkIterator: AtHistogram not implemented"))
+	if i.valType != chunkenc.ValHistogram {
+		panic(fmt.Errorf("chunkIterator: calling AtHistogram when chunk is of different type %v", i.valType))
+	}
+
+	if i.cacheValid {
+		return i.cachedTime, i.cachedHistogram
+	}
+
+	i.cachedTime, i.cachedHistogram = i.it.AtHistogram()
+	i.cacheValid = true
+	return i.cachedTime, i.cachedHistogram
 }
 
 func (i *chunkIterator) AtFloatHistogram() (int64, *histogram.FloatHistogram) {
-	panic(errors.New("chunkIterator: AtFloatHistogram not implemented"))
+	if i.valType == chunkenc.ValHistogram {
+		// use cache as histogram
+		t, h := i.AtHistogram()
+		return t, h.ToFloat()
+	}
+	if i.valType != chunkenc.ValFloatHistogram {
+		panic(fmt.Errorf("chunkIterator: calling AtFloatHistogram when chunk is of different type %v", i.valType))
+	}
+
+	if i.cacheValid {
+		return i.cachedTime, i.cachedFloatHistogram
+	}
+
+	i.cachedTime, i.cachedFloatHistogram = i.it.AtFloatHistogram()
+	i.cacheValid = true
+	return i.cachedTime, i.cachedFloatHistogram
 }
 
 func (i *chunkIterator) AtT() int64 {
-	t, _ := i.At()
-	return t
+	if i.cacheValid {
+		return i.cachedTime
+	}
+	switch i.valType {
+	case chunkenc.ValFloat:
+		t, v := i.At()
+		i.cacheValid = true
+		i.cachedValue = v
+		return t
+	case chunkenc.ValHistogram:
+		t, h := i.AtHistogram()
+		i.cacheValid = true
+		i.cachedHistogram = h
+		return t
+	case chunkenc.ValFloatHistogram:
+		t, fh := i.AtFloatHistogram()
+		i.cacheValid = true
+		i.cachedFloatHistogram = fh
+		return t
+	default:
```

Comment:
eh, actually the `At*` calls already set the cache correctly, no need to set cacheValid and `cache*` values. I'm going to write a test to check the caching

Reply:
i think nothing stops the client from calling only `AtT` without calling `At`, `AtHistogram` or `AtFloatHistogram`. At least that's what I gathered from the godocs of the `chunkenc.Iterator` interface.

You can also not set `cacheValid=true` in this function. And that will also make `AtT` behave consistently.

Reply:
Yes, most of the time AtT will be called without calling At* first, but that's handled by the case where cacheValid==false.

I've removed setting cacheValid=true, however since the functions At(), AtHistogram(), AtFloatHistogram all set it, the unit test I'm adding checks that it's still handled correctly.